calyx/src/analysis/dominator_map.rs

```diff
@@ -0,0 +1,489 @@
+                    hs = hs.union(&stmt_final).copied().collect()
+                }
+            }
+        }
+        hs
+    }
+
+    //Given an id and its predecessors pred, and a domination map d_map, updates
+    //d_map accordingly (i.e. the union of all dominators of the predecessors
+    //plus itself).
+    fn update_node(pred: &HashSet<u64>, id: u64, d_map: &mut DominatorMap) {
+        let mut union: HashSet<u64> = HashSet::new();
+        for id in pred.iter() {
+            if let Some(dominators) = d_map.map.get(id) {
+                union = union.union(dominators).copied().collect();
+            }
+        }
+        union.insert(id);
+        d_map.map.insert(id, union);
+    }
+
+    //Looks through each "node" in the "graph" and updates the dominators accordingly
+    fn update_map(
+        main_c: &ir::Control,
+        cur_id: u64,
+        pred: &HashSet<u64>,
+        d_map: &mut DominatorMap,
+    ) {
+        let c = match Self::get_control(cur_id, main_c) {
+            Some(control) => control,
+            None => return,
+        };
+        match c {
+            ir::Control::Empty(_) => {
+                unreachable!(
+                    "should not pattern match agaisnt empty in update_map()"
+                )
+            }
+            ir::Control::Invoke(_) | ir::Control::Enable(_) => {
+                Self::update_node(pred, cur_id, d_map);
+            }
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                //Could try to think a way of doing it w/o this first stuff
+                let mut first = true;
+                let mut prev_id = cur_id;
+                for stmt in stmts {
+                    let id = Self::get_id(stmt, true);
+                    if first {
+                        Self::update_map(main_c, id, pred, d_map);
+                        first = false;
+                    } else {
+                        Self::update_map(
+                            main_c,
+                            id,
+                            &d_map
+                                .exits_map
+                                .get(&prev_id)
+                                .unwrap_or_else(|| {
+                                    unreachable!(
+                                        "exit node map does not have value for {}",
+                                        prev_id
+                                    )
+                                })
+                                .clone(),
+                            d_map,
+                        );
+                    }
+                    prev_id = Self::get_id(stmt, false);
+                }
+            }
+            ir::Control::Par(ir::Par { stmts, .. }) => {
+                for stmt in stmts {
+                    let id = Self::get_id(stmt, true);
+                    Self::update_map(main_c, id, pred, d_map);
+                }
+            }
+            ir::Control::While(ir::While { body, .. }) => {
```

Comment:
Can you write a comment explaining why it is safe to ignore the comb group in the `while` and `if` cases?

Reply:
I'm not ignoring the comb group-- in the domination map, if I attach a NODE_ID to a while control statement, that "node" refers to only the while guard (which I include to mean its combinational group and its lt.out).

Reply:
Okay, can you add a comment to that effect in the code?